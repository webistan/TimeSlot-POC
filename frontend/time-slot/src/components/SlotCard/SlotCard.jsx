import React, { createContext, useState } from "react";
import { addSlotInObject, deleteSlotData } from "../../redux/action/SlotAction";

import AddSlotPopUp from "../PopUp/AddSlotPopUp";
import _ from "lodash";
import { connect } from "react-redux";

//import moment from "moment";

export const MyContext = createContext("");

//***************************** SlotCard Method **********************************//
function SlotCard(props) {
  const { slotList, addSlotInObject, deleteSlotData } = props;
  const [open, setOpen] = useState(false);
  const [slotsObj, setSlotsObj] = useState({});
  const [slotCopyDay, setSlotCopyDay] = useState([
    {
      day: "Sunday",
      showDay: "S",
      isSelected: false,
      copyDay: false,
    },
    {
      day: "Monday",
      showDay: "M",
      isSelected: false,
      copyDay: false,
    },
    {
      day: "Tuesday",
      showDay: "T",
      isSelected: false,
      copyDay: false,
    },
    {
      day: "Wednesday",
      showDay: "W",
      isSelected: false,
      copyDay: false,
    },
    {
      day: "Thursday",
      showDay: "Th",
      isSelected: false,
      copyDay: false,
    },
    {
      day: "Friday",
      showDay: "F",
      isSelected: false,
      copyDay: false,
    },
    {
      day: "Saturday",
      showDay: "Sat",
      isSelected: false,
      copyDay: false,
    },
  ]);

  //***************************** View Modal Function **********************************//
  const viewModal = async() => {
    console.log('open popup')
   
    let emptyObj = {};
    setSlotsObj((oldState) => ({
      ...emptyObj,
    }));

    let copyArr = JSON.parse(JSON.stringify(slotCopyDay));
    copyArr.forEach((item, idx) => {
      if (item.day === props.keys[0]) {
        item["isSelected"] = true;
      }
    });
    setSlotCopyDay(copyArr);

    setOpen(true);
  };

  //***************************** PopUp Function **********************************//
  const viewPopUp = (day) => {
    return (
      <>
        <MyContext.Provider
          value={{
            addSlotsObj,
            day,
            open,
            onClose,
            slotsObj,
            slotList,
            slotCopyDay,
            copySolts,
            removeSlotsObj,
          }}
        >
          <AddSlotPopUp />
        </MyContext.Provider>
      </>
    );
  };

  //***************************** Add Slots in PopUp **********************************//
  const addSlotsObj = (data) => {
    console.log("addslotdata", data);
    setSlotsObj((oldState) => ({
      ...data,
    }));
  };

  //***************************** Remove Slots in PopUp **********************************//
  const removeSlotsObj = (data) => {
    let idx = data.index;
    let day = data.item.allocated_day;
    console.log("day1", day, idx);

    let copySlotObj = JSON.parse(JSON.stringify(slotsObj));

    let json = {};
    copySlotObj &&
      Object.keys(copySlotObj).forEach(function (key) {
        copySlotObj[key].map((item, index) => {
          if (idx === index) {
            copySlotObj[key].splice(index, 1);
          }
          let jsonSlot = {
            [day]: copySlotObj[key] ? copySlotObj[key] : [],
          };
          Object.assign(json, jsonSlot);
        });
      });
      if (json[day] && json[day].length === 0) {
        let emptyObj = {};
        setSlotsObj((oldState) => ({
          ...emptyObj,
        }));
      } else {
        setSlotsObj((oldState) => ({
          ...json,
        }));
      }
  };

  //***************************** Copy Slots method **********************************//
  const copySlotDataInFinalObj = async (
    newJsonData,
    slotCopyDay,
    newSlotObj
  ) => {
    let allUpdatedData = JSON.parse(JSON.stringify(newJsonData));
    let selectCopyDay = _.filter(slotCopyDay, { copyDay: true });

    let obj = {};
    if (selectCopyDay && selectCopyDay.length > 0) {
      _.forEach(selectCopyDay, async function (item, idx) {
        let arr = [];
        _.forEach(newSlotObj[props.keys[0]], async function (value, index) {
          arr.push({
            start_time: value.start_time,
            allocated_slot: value.allocated_slot,
            allocated_day: item.day,
          });
          obj[item.day] = arr;
        });
      });

      if (!_.isEmpty(obj)) {
        Object.keys(obj).map((value) => {
          if (
            allUpdatedData["slots"][value] &&
            allUpdatedData["slots"][value].length > 0
          ) {
            let newArr = allUpdatedData["slots"][value];
            const merged = [...newArr, ...obj[value]];
            allUpdatedData["slots"][value] = merged;
          } else {
            allUpdatedData["slots"][value] = obj[value];
          }
        });
      }
    }

    return allUpdatedData;
  };

  //***************************** Close PopUp **********************************//
  const onClose = async (data) => {
    console.log("dataa areee for popup close:: ", data);
    if (data && data.popUpClose && !data.addData) {
      setOpen(false);
      let copyArr1 = JSON.parse(JSON.stringify(slotCopyDay));
      copyArr1.forEach((item, idx) => {
           item["isSelected"] = false;
           item['copyDay'] = false
      });
      console.log('copyArr1 are:: ',copyArr1)
      setSlotCopyDay(copyArr1);
    }

    if (data && data.popUpClose && data.addData) {
      setOpen(false);
      setSlotsObj((oldState) => ({
        ...data.obj,
      }));

      //console.log("slotsObj in card page:: ", slotsObj);
      const newSlotObj = JSON.parse(JSON.stringify(slotsObj));
      //console.log("newSlotObject", JSON.stringify(newSlotObj));

      let newJsonData = JSON.parse(JSON.stringify(slotList));
      //console.log("newJsonData", JSON.stringify(newJsonData));

      let addSlotArr = JSON.parse(JSON.stringify(newSlotObj[data.selectedDay]));
      //console.log('addSlotArr are:::: ',JSON.stringify(addSlotArr))

      let destinationSlotArr = JSON.parse(
        JSON.stringify(newJsonData["slots"][props.keys[0]])
      );
      //console.log('destinationSlotArr are:::: ',destinationSlotArr)

      // Merge arrays
      const merged = [...destinationSlotArr, ...addSlotArr];

      //console.log('merged are:: ',merged)

      newJsonData.slots[props.keys[0]] = merged;
      //console.log('slotCopyDay data are::: ', JSON.stringify(slotCopyDay))

      //console.log('final newJsonData are:: ',JSON.stringify(newJsonData))

      let finalObj = await copySlotDataInFinalObj(
        newJsonData,
        slotCopyDay,
        newSlotObj
      );
      console.log("finalObj areeeeeeee@@@@@@@:: ", finalObj);
      console.log("doneeee");
      addSlotInObject(finalObj);
    }
  };
  const copySolts = (day) => {
    let copyArr = JSON.parse(JSON.stringify(slotCopyDay));
    copyArr.forEach((item, idx) => {
      if (item.day === day) {
        item["isSelected"] = true;
        item["copyDay"] = true;
      }
    });
    setSlotCopyDay(copyArr);
  };

  //***************************** Function Of Delete Slots In SlotCard **********************************//
  const deleteSlot = (start_time, allocated_day) => {
    let start_date = slotList && slotList.start_date;
    let end_date = slotList && slotList.end_date;
    deleteSlotData(start_time, allocated_day, start_date, end_date);
  };
  return (
    <>
      <div className="wt-card" data-testid="wt-card">
        <div className="wt-card-head">
          <div className="card-head-left">
            <input type="checkbox" />
            {props.keys && props.keys[0]}
          </div>
          <div className="card-head-right">
            <button className="button1">Copy Schedule</button>
            <button
              disabled={open === true ? true : false}
              className="button2"
              onClick={(e) => {
                viewModal();
              }}
            >
              + Add Slot
            </button>
          </div>
        </div>
        <div className="wt-card-content">
          {props.keys &&
            props.keys[1] &&
            props.keys[1].length > 0 &&
            props.keys[1].map((value, idx) => (
              <div className="wt-block" key={idx}>
                <div className="input-group">
                  {/* <input type="time" className="w-130" defaultValue={moment(value.slot_time, 'hh:mm A').format('HH:mm:ss')}/> */}
                  <input
                    type="time"
                    className="w-130"
                    defaultValue={value.start_time}
                  />
                  <input
                    type="text"
                    className="w-90"
                    defaultValue={value.allocated_slot}
                  />
                  {value && value.id ? (
                    <span
                      className="delete-btn"
                      onClick={() =>
                        deleteSlot(value.start_time, value.allocated_day)
                      }
                    >
                      <i className="fa-sharp fa-solid fa-trash" />
                    </span>
                  ) : null}
                </div>
              </div>
            ))}
        </div>
        {open === true ? viewPopUp(props.keys[0]) : null}
      </div>
    </>
  );
}

//***************************** MapStateToProps Method **********************************//
const mapStateToProps = (state) => ({
  slotList: state.slotReducer.slotList,
});

//***************************** MapDispatchToProps Method **********************************//
const mapDispatchToProps = {
  addSlotInObject,
  deleteSlotData,
};

export default connect(mapStateToProps, mapDispatchToProps)(SlotCard);
