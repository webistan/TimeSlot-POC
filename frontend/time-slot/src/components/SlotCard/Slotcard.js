const SlotCard = (props) => {
    const { slots } = props
    // data-testid is a testing id 
    // which is used only during tests
    return (
        <>

        
        <div data-testid="Slot-Card-1"  className="wt-card">
            <div className="wt-card-head">
            <div className="card-head-left"><input type="checkbox" /> Monday</div>
            <div className="card-head-right">
                <button className="button1">Copy Schedule</button>
                <button className="button2">+ Add Slot</button>
            </div>
            </div>
            <div className="wt-card-content">
            {/* slot:start*/}
            <div className="wt-block">
                <div className="input-group">
                <input type="time" className="w-130" />
                <input data-testid="slot-box" type="text" className="w-90" defaultValue={"10"}/>
                <span className="delete-btn"><i className="fa-sharp fa-solid fa-trash" /></span>
                </div>
            </div>
            {/* slot:end*/}
            {/* slot:start*/}
            <div className="wt-block">
                <div className="input-group">
                <input type="time" className="w-130" />
                <input type="text" className="w-90" />
                <span className="delete-btn"><i className="fa-sharp fa-solid fa-trash" /></span>
                </div>
            </div>
            {/* slot:end*/}
            </div>
        </div>
      
        </>
    )
}
  
export default SlotCard;