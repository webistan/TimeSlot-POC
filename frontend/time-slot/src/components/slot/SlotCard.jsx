import React, { createContext, useState } from "react";
import { addSlotInObject, deleteSlotData } from "../../redux/action/SlotAction";

import AddSlotPopUp from "./AddSlotPopUp";
import { connect } from "react-redux";
import moment from "moment";

export const MyContext = createContext();

function SlotCard(props) {
  const { slotList, addSlotInObject ,deleteSlotData} = props;
  const [open, setOpen] = useState(false);
  const [slotsObj, setSlotsObj] = useState({});

  const timeSlotJson = {
    start_date: "",

    end_date: "",

    slots: {
      Tuesday: [
        {
          start_time: "02:30PM",

          allocated_slot: 10,
        },
      ],

      Monday: [
        {
          start_time: "09:30AM",

          allocated_slot: 10,
        },
      ],
    },
  };

  // const [popUpOpen, setPopUpOpen] = useState(false);
  console.log("props", props.keys[1]);

  const val = props.keys[1].map((value, idx) => {
    console.log("value098", value.slot_time);
    return value;
  });
  console.log("val123", val);

  const viewModal = () => {
    let currentDayData = props.keys[1].filter(
      (o) => o.allocated_day === props.keys[0]
    );
    console.log("currentDayData aree:: ", currentDayData);
    let json = {};
    json[props.keys[0]] = currentDayData;
    setSlotsObj((oldState) => ({
      ...json,
    }));
    setOpen(true);
  };

  const viewPopUp = (day) => {
    console.log("day are:: ", day);

    return (
      <>
        <MyContext.Provider
          value={{ addSlotsObj, day, open, onClose, slotsObj, slotList }}
        >
          <AddSlotPopUp />
        </MyContext.Provider>
      </>
    );
  };
  const closeViewPopUp = () => {
    setOpen(false);
  };

  const addSlotsObj = (data) => {
    setSlotsObj((oldState) => ({
      ...data,
    }));
  };

  const onClose = (data) => {
    console.log("dataa areee for popup close:: ", data);
    if (data && data.popUpClose && !data.addData) {
      console.log("close popup data reset");
      setOpen(false);
    }

    if (data && data.popUpClose && data.addData) {
      console.log("close popup data added");
      setOpen(false);
      setSlotsObj((oldState) => ({
        ...data.obj,
      }));

      console.log("slotsObj in card page:: ", slotsObj);
      const newSlotObj = JSON.parse(JSON.stringify(slotsObj));
      console.log("newSlotObject", newSlotObj);

      // let datafiltered = Object.entries(newSlotObj).filter((o) => o);
      // const filterValue = datafiltered[0].filter((value) => value);
      // const jsonData = filterValue[1].filter((value) => !value.id);
      // console.log("datvalue1", jsonData);

      let newJsonData = JSON.parse(JSON.stringify(slotList));
      console.log("newJsonData", newJsonData);

      newJsonData["start_date"] = "20-09-2022";
      newJsonData["end_date"] = "20-11-2022";
      newJsonData["slots"][props.keys[0]] = newSlotObj[props.keys[0]];
      console.log("newJsonData123", newJsonData);
      addSlotInObject(newJsonData);
    }
  };

  console.log("slotList are:: ", slotList);

  const deleteSlot=(start_time,allocated_day)=>{
    console.log("delete",start_time,allocated_day)
    deleteSlotData(start_time,allocated_day)

  }

  return (
    <>
      <div className="wt-card">
        <div className="wt-card-head">
          <div className="card-head-left">
            <input type="checkbox" />
            {props.keys[0]}
          </div>
          <div className="card-head-right">
            <button className="button1">Copy Schedule</button>
            <button
              className="button2"
              onClick={(e) => {
                viewModal();
              }}
            >
              + Add Slot
            </button>
          </div>
        </div>
        <div className="wt-card-content">
          {props.keys &&
            props.keys[1] &&
            props.keys[1].length > 0 &&
            props.keys[1].map((value, idx) => (
              <div className="wt-block" key={idx}>
                <div className="input-group">
                  {/* <input type="time" className="w-130" defaultValue={moment(value.slot_time, 'hh:mm A').format('HH:mm:ss')}/> */}
                  <input
                    type="time"
                    className="w-130"
                    defaultValue={value.start_time}
                  />
                  <input
                    type="text"
                    className="w-90"
                    defaultValue={value.allocated_slot}
                  />
                  <span
                    className="delete-btn"
                    onClick={() =>
                      deleteSlot(value.start_time, value.allocated_day)
                    }
                  >
                    <i className="fa-sharp fa-solid fa-trash" />
                  </span>
                </div>
              </div>
            ))}
        </div>
        {open === true ? viewPopUp(props.keys[0]) : null}

        {/* <div className="wt-card-content">
          <div className="wt-block">
            <div className="input-group">
              <input type="time" className="w-130"/>
              <input type="text" className="w-90" defaultValue={10} />
              <span className="delete-btn"><i className="fa-sharp fa-solid fa-trash" /></span>
            </div>
          </div>
        </div> */}
      </div>
    </>
  );
}

const mapStateToProps = (state) => ({
  slotList: state.slotReducer.slotList,
});

const mapDispatchToProps = {
  addSlotInObject,
  deleteSlotData
};

export default connect(mapStateToProps, mapDispatchToProps)(SlotCard);
