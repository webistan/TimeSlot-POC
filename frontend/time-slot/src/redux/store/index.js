import { applyMiddleware, createStore } from 'redux'

import { createLogger } from 'redux-logger'
import createSagaMiddleware from 'redux-saga'
import rootReducer from '../../redux/reducer'
import sagas from '../../redux/saga'

// import sagas from './redux/saga';
// import rootReducer from './redux/reducer';

const loggerMiddleware = createLogger()

const sagaMiddleware = createSagaMiddleware()

const store = createStore(rootReducer, applyMiddleware(sagaMiddleware,loggerMiddleware))

sagaMiddleware.run(sagas)

export default store

