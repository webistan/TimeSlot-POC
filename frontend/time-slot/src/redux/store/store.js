import { configureStore } from "@reduxjs/toolkit";
import slot from "../reducer/index";

export const createStore = () =>
  configureStore({
    reducer: {
      slotReducer: slot,
    },
  });
