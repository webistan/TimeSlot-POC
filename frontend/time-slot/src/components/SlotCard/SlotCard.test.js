// Importing the jest testing library
import "@testing-library/jest-dom";

import { cleanup, render } from "@testing-library/react";

import { Provider } from "react-redux";
import SlotCard from "./SlotCard";
import { createStore } from "../../redux/store/store";
import renderer from "react-test-renderer";

// afterEach function runs after each test suite is executed
afterEach(() => {
  cleanup(); // Resets the DOM after each test suite
});

describe("SlotCard Component", () => {
  test("Should check card be exist in Component", () => {
    const { getByTestId } = render(
      <Provider store={createStore()}>
        <SlotCard />
      </Provider>
    );
    const card = getByTestId("wt-card");
    expect(card).toBeInTheDocument();

    const SlotCardDomTree = renderer
      .create(
        <Provider store={createStore()}>
          <SlotCard />
        </Provider>
      )
      .toJSON();
    expect(SlotCardDomTree).toMatchSnapshot();
  });
});
