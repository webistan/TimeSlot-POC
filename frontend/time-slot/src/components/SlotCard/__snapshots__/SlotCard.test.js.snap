// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SlotCard Component Should check card be exist in Component 1`] = `
<div
  className="wt-card"
  data-testid="wt-card"
>
  <div
    className="wt-card-head"
  >
    <div
      className="card-head-left"
    >
      <input
        type="checkbox"
      />
    </div>
    <div
      className="card-head-right"
    >
      <button
        className="button1"
      >
        Copy Schedule
      </button>
      <button
        className="button2"
        disabled={true}
        onClick={[Function]}
      >
        + Add Slot
      </button>
    </div>
  </div>
  <div
    className="wt-card-content"
  />
</div>
`;
